File: android/app/src/main/java/com/bensbible/app/model/VerseID.kt
package com.bensbible.app.model

data class VerseID(
    val book: String,
    val chapter: Int,
    val verse: Int
) : Comparable<VerseID> {

    val key: String get() = "$book:$chapter:$verse"
    val displayReference: String get() = "$book $chapter:$verse"

    override fun toString(): String = key

    override fun compareTo(other: VerseID): Int {
        val bookCmp = book.compareTo(other.book)
        if (bookCmp != 0) return bookCmp
        val chapterCmp = chapter.compareTo(other.chapter)
        if (chapterCmp != 0) return chapterCmp
        return verse.compareTo(other.verse)
    }

    companion object {
        fun fromKey(key: String): VerseID? {
            val parts = key.split(":")
            if (parts.size != 3) return null
            val chapter = parts[1].toIntOrNull() ?: return null
            val verse = parts[2].toIntOrNull() ?: return null
            return VerseID(book = parts[0], chapter = chapter, verse = verse)
        }

        fun displayRange(verseIDs: Set<VerseID>): String {
            if (verseIDs.isEmpty()) return ""
            val sorted = verseIDs.sorted()
            val first = sorted.first()
            val book = first.book
            val chapter = first.chapter

            val verses = sorted.map { it.verse }
            val ranges = mutableListOf<String>()
            var rangeStart = verses[0]
            var rangeEnd = verses[0]

            for (i in 1 until verses.size) {
                if (verses[i] == rangeEnd + 1) {
                    rangeEnd = verses[i]
                } else {
                    ranges.add(if (rangeStart == rangeEnd) "$rangeStart" else "$rangeStart-$rangeEnd")
                    rangeStart = verses[i]
                    rangeEnd = verses[i]
                }
            }
            ranges.add(if (rangeStart == rangeEnd) "$rangeStart" else "$rangeStart-$rangeEnd")

            return "$book $chapter:${ranges.joinToString(", ")}"
        }
    }
}

File: BensBible2/Models/VerseID.swift
import Foundation

struct VerseID: Hashable, Equatable, CustomStringConvertible {
    let book: String
    let chapter: Int
    let verse: Int

    var key: String { "\(book):\(chapter):\(verse)" }
    var description: String { key }

    var displayReference: String {
        "\(book) \(chapter):\(verse)"
    }

    init(book: String, chapter: Int, verse: Int) {
        self.book = book
        self.chapter = chapter
        self.verse = verse
    }

    init?(key: String) {
        let parts = key.split(separator: ":")
        guard parts.count == 3,
              let chapter = Int(parts[1]),
              let verse = Int(parts[2]) else { return nil }
        self.book = String(parts[0])
        self.chapter = chapter
        self.verse = verse
    }

    static func displayRange(from verseIDs: Set<VerseID>) -> String {
        guard let first = verseIDs.first else { return "" }

        let book = first.book
        let chapter = first.chapter
        let verses = verseIDs.map(\.verse).sorted()

        var ranges: [String] = []
        var rangeStart = verses[0]
        var rangeEnd = verses[0]

        for i in 1..<verses.count {
            if verses[i] == rangeEnd + 1 {
                rangeEnd = verses[i]
            } else {
                ranges.append(rangeStart == rangeEnd ? "\(rangeStart)" : "\(rangeStart)-\(rangeEnd)")
                rangeStart = verses[i]
                rangeEnd = verses[i]
            }
        }
        ranges.append(rangeStart == rangeEnd ? "\(rangeStart)" : "\(rangeStart)-\(rangeEnd)")

        return "\(book) \(chapter):\(ranges.joined(separator: ", "))"
    }
}
